pyterrier/transformer.py

```diff
@@ -93,16 +101,28 @@ def transform(self, topics_or_res : pd.DataFrame) -> pd.DataFrame:
             Abstract method for all transformations. Typically takes as input a Pandas
             DataFrame, and also returns one.
         """
-        pass
+        # We should have no recursive transform <-> transform_iter problem, due to the __new__ check, UNLESS .transform() is called on an Indexer.
+        return pd.DataFrame(list(self.transform_iter(topics_or_res.to_dict(orient='records'))))
 
-    def transform_iter(self, input: Iterable[dict]) -> pd.DataFrame:
+    def transform_iter(self, input: Iterable[dict]) -> Iterable[dict]:
         """
-            Method that proesses an iter-dict by instantiating it as a dataframe and calling transform().
-            Returns the DataFrame returned by transform(). This can be a handier version of transform()
-            that avoids constructing a dataframe by hand. Alo used in the implementation of index() on a composed 
-            pipeline.
+            Method that proesses an iter-dict by instantiating it as a dataframe and calling ``transform()``.
+            Returns an Iterable[dict] equivalent to the DataFrame returned by ``transform()``, usually a list or a generator. 
+            This can be a handier version of ``transform()`` that avoids constructing a dataframe. Also used in the 
+            instantiation of ``index()`` on a composed pipeline.
         """
-        return self.transform(pd.DataFrame(list(input)))
+        # We should have no recursive transform <-> transform_iter problem, due to the __new__ check, UNLESS .transform() is called on an Indexer.
+        return iter(self.transform(pd.DataFrame(list(input))).to_dict(orient='records'))
+    
+    def __call__(self, input : Union[pd.DataFrame, Iterable[dict]]) -> Union[pd.DataFrame, Iterable[dict]]:
+        """
+            Sets up a default method for every transformer, which is aliased to ``transform()`` (for DataFrames)
+            or ``list(transform_iter())`` (for iterable dictionaries) depending on the type of input. The return type
+            matches the input type, but is always instantiated (generators should not be returned).
+        """
+        if isinstance(input, pd.DataFrame):
+            return self.transform(input)
+        return list(self.transform_iter(input))
```

Comment:
I'd rather have `__call__` return whatever type you pass in. So if it's a dataframe, return as a dataframe, if it's a list, return it as a list, if it's a generic iterable, return a generic iterable.

Reply:
Not sure I'm as keen. `.__call__()` for me is user-facing magic for ease-of-use; anything trying to be clever that knows the type should call the correct `.transform()` or `.transform_iter()`? What would the sketch code look like?

Reply:
To me, there's an appeal to having a function that gives out whatever type is put in. If a user is savvy enough to use iterators, they hopefully know how they work.

> What would the sketch code look like?

```python
        if isinstance(input, pd.DataFrame):
            return self.transform(input)
        if isinstance(input, list):
            return list(self.transform_iter(input))
        # Are there other input types we want to handle specifically?
        return self.transform_iter(input)
```

Reply:
Ok. Do you want to make the change (and to the types)?

Reply:
i.e return type of __call__ should now be:
```Union[pd.DataFrame, List[dict], Iterator[dict]]```
?

Also:
 - `dict` vs `Dict` vs `Dict[str,Any]`